### DISCUSSION TITLE
psycopg3 on AWS Lambda

### DISCUSSION THREAD

**Discussion by kepee24** [psycopg/psycopg#319] [Q&A]
Hi Team,

I have used psycopg2 earlier on AWS Lambda, to automate some of our Postgres archival workflows.
Wanted to know if its possible to use psycopg3 as well? Currently, unable to find a download link for this release. 
Use case is:
Copy some data from a table hosted in one Postgres DB to another table on another Postgres DB.

**Comment by HacKanCuBa** [comment]
It is currently not possible, due to this lib being dynamically linked to `libpq`, which is not available at runtime in AWS Lambda.
We would need to compile and statically link said lib.
There's a [project for that for psycopg2](https://github.com/AbhimanyuHK/aws-psycopg2).

**Comment by aliavni** [comment]
Hi everyone, do you have any suggestions on how to use psycopg3 in AWS Lambda? I get the following error when I install `psycopg[binary]==3.1.12` and package it for lambda 


```
Traceback (most recent call last):
  File "/var/task/mangum/protocols/http.py", line 58, in run
    await app(self.scope, self.receive, self.send)
  File "/var/task/fastapi/applications.py", line 292, in __call__
    await super().__call__(scope, receive, send)
  File "/var/task/starlette/applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/var/task/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/var/task/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/var/task/starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "/var/task/starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "/var/task/fastapi/middleware/asyncexitstack.py", line 20, in __call__
    raise e
  File "/var/task/fastapi/middleware/asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
  File "/var/task/starlette/routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "/var/task/starlette/routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "/var/task/starlette/routing.py", line 66, in app
    response = await func(request)
               ^^^^^^^^^^^^^^^^^^^
  File "/var/task/fastapi/routing.py", line 263, in app
    solved_result = await solve_dependencies(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/fastapi/dependencies/utils.py", line 580, in solve_dependencies
    solved = await solve_generator(
             ^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/fastapi/dependencies/utils.py", line 505, in solve_generator
    return await stack.enter_async_context(cm)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lang/lib/python3.11/contextlib.py", line 638, in enter_async_context
    result = await _enter(cm)
             ^^^^^^^^^^^^^^^^
  File "/var/lang/lib/python3.11/contextlib.py", line 204, in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/app/db.py", line 26, in get_session
    engine = get_engine(client)
             ^^^^^^^^^^^^^^^^^^
  File "/var/task/app/db.py", line 14, in get_engine
    engine = AsyncEngine(create_engine(db_url, echo=True, future=True))
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 2, in create_engine
  File "/var/task/sqlalchemy/util/deprecations.py", line 281, in warned
    return fn(*args, **kwargs)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^
  File "/var/task/sqlalchemy/engine/create.py", line 601, in create_engine
    dbapi = dbapi_meth(**dbapi_args)
            ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/sqlalchemy/dialects/postgresql/psycopg.py", line 706, in import_dbapi
    import psycopg
  File "/var/task/psycopg/__init__.py", line 9, in <module>
    from . import pq  # noqa: F401 import early to stabilize side effects
    ^^^^^^^^^^^^^^^^
  File "/var/task/psycopg/pq/__init__.py", line 114, in <module>
    import_from_libpq()
  File "/var/task/psycopg/pq/__init__.py", line 106, in import_from_libpq
    raise ImportError(
ImportError: no pq wrapper available.
Attempts made:
- couldn't import psycopg 'c' implementation: No module named 'psycopg_c'
- couldn't import psycopg 'binary' implementation: No module named 'psycopg_binary'
- couldn't import psycopg 'python' implementation: libpq library not foundEND RequestId: e1e82ef6-98f9-4440-9ed2-e3a987fea359
```

**Reply by HacKanCuBa** [reply]
short answer: you can't, given lambda env doesn't have required libs
long answer: you can push a [custom image](https://docs.aws.amazon.com/lambda/latest/dg/images-create.html) to lambda, instead of using the lambda environment

**Reply by aliavni** [reply]
thank you

**Comment by FinchPowers** [comment]
Hi!

I was able to run psycopg 3.2.1 on an AWS Lambda with the python 3.11 runtime, x86_64 architecture. I installed it with the binary extra.

The problem I had that lead me here is that when trying to create the connection through [sqlalchemy](https://www.sqlalchemy.org/), it errored out with `psycopg version 3.0.2 or higher is required.`. I was bundling with [serverless-python-requirements](https://github.com/serverless/serverless-python-requirements/) with the settings `slim: true`, `strip: false`. Strip false is important to keep the `.so` files. The problem was with `slim: true`, which DID NOT ship the `psycopg-3.2.1.dist-info` directory. This is what is being used by psycopg to determine its own version. 

Try this twice: once with `site-packages/psycopg-3.2.1.dist-info` and once without.

```python
import psycopg
print(psycopg.version.__version__)
```

With it, it correctly prints `3.2.1`.
Without it, it prints `0.0.0.0`.

**Reply by jaypinho** [reply]
@FinchPowers do you mind sharing how you successfully installed and ran psycopg 3+ on Lambda? I've spent much of today trying and failing to do so:

Local env: macOS (M1)
Python 3.13
Lambda architecture: x86-64

Depending on which variations of installations I try, I get error messages like the one about not finding a pq wrapper, or not finding psycopg (c, binary, or Python), etc.

Thank you!

**Reply by FinchPowers** [reply]
I use the following stack
* servlerless 3
  * plugin serverless-python-requirements
    * It's important to configure it with `slim: false`, or it will strip a required file. (If you get it to work you can toggle it back on and make sure to just keep the right files, but to debug it's easier to go with `false`. I can help you back when you get there.)
* poetry
* python 3.12
  * psycopg 3.2.3
  * psycopg-binary 3.2.3
* Local env: macOS (M1)
* Lambda architecture: x86-64

I can't think of any other relevant detail. If you want more info, ask!

**Reply by jaypinho** [reply]
Thank you!!

**Reply by AzySir** [reply]
@FinchPowers were the packages installed locally on your lambda - or in lambda layers?

**Reply by FinchPowers** [reply]
@AzySir Directly in the lambda, no layer.

**Reply by AzySir** [reply]
Thanks for such a quick reply @FinchPowers - If that's my last resort i will implement it that way - i can't believe the lack of integration between AWS and this .. frustrating stuff!

**Reply by FinchPowers** [reply]
@AzySir I'm not certain, but I think the problem has more to do with psycopg3 (where it stores the version number) and how we package it to run in AWS lambda. (We often strip that location to make the package smaller.)

**Reply by dvarrazzo** [reply]
@FinchPowers I don't know who you refer to in "how we package it" (and, for everyone's information, that doesn't include myself), but, if you are creating some psycopg-derived package and stripping the packaging metadata, you might want to inject the literal of the version in the package you are creating:

https://github.com/psycopg/psycopg/blob/1e7a01667210a83979caa116b844f58dea5ee5e8/psycopg/psycopg/version.py#L9-L12

Just the moral equivalent of `sed "s/0\.0\.0\.0/$PACKAGE_VERSION/"` on this file (and similar `version.py` files in other psycopg packages) should do.

**Reply by AzySir** [reply]
> @AzySir I'm not certain, but I think the problem has more to do with psycopg3 (where it stores the version number) and how we package it to run in AWS lambda. (We often strip that location to make the package smaller.)

Thanks I've elected to use another package - I can't understand why AWS recommend this framework if the framework doesn't meet the required libraries to operate in Lambda.
